perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+  }
+  RCLCPP_INFO(rclcpp::get_logger("BEVDet"), "%s", oss.str().c_str());
+}
+
+int BEVDet::initEngine(const std::string & engine_file)
+{
+  if (deserializeTRTEngine(engine_file, &trt_engine_)) {
+    return EXIT_FAILURE;
+  }
+
+  if (trt_engine_ == nullptr) {
+    RCLCPP_ERROR(rclcpp::get_logger("BEVDet"), "Failed to deserialize engine file!");
+    return EXIT_FAILURE;
+  }
+  trt_context_ = trt_engine_->createExecutionContext();
+
+  if (trt_context_ == nullptr) {
+    RCLCPP_ERROR(rclcpp::get_logger("BEVDet"), "Failed to create TensorRT Execution Context!");
+    return EXIT_FAILURE;
+  }
+
+  // set bindings
+  std::vector<nvinfer1::Dims32> shapes{
+    {4, {n_img_, 3, src_img_h_, src_img_w_ / 4}},
+    {1, {3}},
+    {1, {3}},
+    {3, {1, n_img_, cam_params_size_}},
+    {1, {valid_feat_num_}},
+    {1, {valid_feat_num_}},
+    {1, {valid_feat_num_}},
+    {1, {unique_bev_num_}},
+    {1, {unique_bev_num_}},
+    {5, {1, adj_num_, bevpool_channel_, bev_h_, bev_w_}},
+    {3, {1, adj_num_, transform_size_}},
+    {2, {1, 1}}};
+
+  for (size_t i = 0; i < shapes.size(); i++) {
+    trt_context_->setBindingDimensions(i, shapes[i]);
+  }
+
+  buffer_map_.clear();
+  for (auto i = 0; i < trt_engine_->getNbBindings(); i++) {
+    auto dim = trt_context_->getBindingDimensions(i);
+    auto name = trt_engine_->getBindingName(i);
+    buffer_map_[name] = i;
+    print_dim(dim, name);
+  }
+
+  return EXIT_SUCCESS;
+}
+
+int BEVDet::deserializeTRTEngine(
+  const std::string & engine_file, nvinfer1::ICudaEngine ** engine_ptr)
+{
+  std::stringstream engine_stream;
+  engine_stream.seekg(0, engine_stream.beg);
+
+  std::ifstream file(engine_file);
+  engine_stream << file.rdbuf();
+  file.close();
+
+  nvinfer1::IRuntime * runtime = nvinfer1::createInferRuntime(g_logger_);
+  if (runtime == nullptr) {
+    std::string msg("Failed to build runtime parser!");
+    g_logger_.log(nvinfer1::ILogger::Severity::kERROR, msg.c_str());
+    return EXIT_FAILURE;
+  }
+  engine_stream.seekg(0, std::ios::end);
+  const int engine_size = engine_stream.tellg();
+
+  engine_stream.seekg(0, std::ios::beg);
+  void * engine_str = malloc(engine_size);
+  engine_stream.read(reinterpret_cast<char *>(engine_str), engine_size);
+
+  nvinfer1::ICudaEngine * engine = runtime->deserializeCudaEngine(engine_str, engine_size, NULL);
+  if (engine == nullptr) {
+    std::string msg("Failed to build engine parser!");
+    g_logger_.log(nvinfer1::ILogger::Severity::kERROR, msg.c_str());
+    return EXIT_FAILURE;
+  }
+  *engine_ptr = engine;
+  for (int bi = 0; bi < engine->getNbBindings(); bi++) {
+    if (engine->bindingIsInput(bi) == true) {
+      RCLCPP_INFO(
+        rclcpp::get_logger("BEVDet"), "Binding %d (%s): Input.", bi, engine->getBindingName(bi));
+    } else {
+      RCLCPP_INFO(
+        rclcpp::get_logger("BEVDet"), "Binding %d (%s): Output.", bi, engine->getBindingName(bi));
+    }
+  }
+  return EXIT_SUCCESS;
+}
+
+void BEVDet::getAdjBEVFeature(
```

Comment:
Namespaces: Wrap the function `getAdjBEVFeature` within the  `autoware_tensorrt_bevdet` namespace.